terraform/interpolate.go

```diff
@@ -589,21 +590,19 @@ func (i *Interpolater) interpolateComplexTypeAttribute(
 			return unknownVariable(), nil
 		}
 
-		var keys []string
+		keys := make([]string, 0)
 		listElementKey := regexp.MustCompile("^" + resourceID + "\\.[0-9]+$")
 		for id, _ := range attributes {
 			if listElementKey.MatchString(id) {
 				keys = append(keys, id)
 			}
 		}
+		sort.Strings(keys)
 
 		var members []string
 		for _, key := range keys {
 			members = append(members, attributes[key])
 		}
-		// This behaviour still seems very broken to me... it retains BC but is
-		// probably going to cause problems in future
-		sort.Strings(members)
```

Comment:
We just discussed this in chat - sorting by keys rather than members here is strictly better, and a major release is definitely the time to make this change.

In order to address the BC issues here of list interpolations unexpectedly shuffling on upgrade, we're going to introduce a `sort()` interpolation function to sort the contents of a list. This should allow users seeing unexpected diffs due to list shuffling on an 0.7 upgrade to retain the old behavior.